src/main/java/com/linecorp/armeria/client/tracing/TracingRemoteInvoker.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.tracing;
+
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+import com.github.kristofa.brave.Brave;
+import com.github.kristofa.brave.ClientRequestAdapter;
+import com.github.kristofa.brave.ClientResponseAdapter;
+import com.github.kristofa.brave.KeyValueAnnotation;
+import com.github.kristofa.brave.SpanId;
+import com.twitter.zipkin.gen.Span;
+
+import com.linecorp.armeria.client.ClientCodec;
+import com.linecorp.armeria.client.ClientOptions;
+import com.linecorp.armeria.client.DecoratingRemoteInvoker;
+import com.linecorp.armeria.client.RemoteInvoker;
+
+import io.netty.util.concurrent.Future;
+
+/**
+ * An abstract {@link RemoteInvoker} that traces remote service invocations.
+ * <p/>
+ * This class depends on <a href="https://github.com/openzipkin/brave">brave</a> distributed tracing library.
+ */
+public abstract class TracingRemoteInvoker extends DecoratingRemoteInvoker {
+
+    private final ClientTracingInterceptor clientInterceptor;
+
+    public TracingRemoteInvoker(RemoteInvoker remoteInvoker, Brave brave) {
+        super(remoteInvoker);
+        clientInterceptor = new ClientTracingInterceptor(brave);
+    }
+
+    @Override
+    public <T> Future<T> invoke(URI uri,
+                                ClientOptions options,
+                                ClientCodec codec,
+                                Method method,
+                                Object[] args) throws Exception {
+
+        // create new request adapter to catch generated spanId
+        final InternalClientRequestAdapter requestAdapter = new InternalClientRequestAdapter(method.getName());
+
+        final Span span = clientInterceptor.openSpan(requestAdapter);
+
+        // new client options with trace data
+        final ClientOptions traceAwareOptions = putTraceData(options, requestAdapter.getSpanId());
```

Comment:
brave's HTTP subproject intends to set "X-B3-Sampled: 0" even if trace is disabled. I'd like to follow the brave's spec for interoperability.

Note: 
Whether to set the header or not affects server-side behaviors.
If there is "X-B3-Sampled: 0", then server side brave interceptor doesn't send spans to zipkin, and set the same header in further RPC call recursively.
If there is not "X-B3-Sampled" header, server would start new tracing (depending on its sampling rate)